### ISSUE TITLE
Issue with Mapped Types' Inferred Types

### ISSUE THREAD

**Issue by mgreystone** [sinclairzx81/typebox#1103]
I have the following generic function that transforms an object schema:

```ts
import { Type, type TSchema, type TObject } from '@sinclair/typebox'

function Nullable<T extends TSchema>(TVal: T) {
  return Type.Union([TVal, Type.Null()])
}

function OptionalNullable<T extends TSchema>(TVal: T) {
  return Type.Optional(Nullable(TVal));
}

export function PartialNullable<T extends TObject>(TObj: T) {
  return Type.Mapped(Type.KeyOf(TObj), (K) =>
    OptionalNullable(Type.Index(TObj, K)),
  );
}
```

This works up until v0.34.1. Typescript no longer compiles this & gives the following error:

> index.ts(11,17): error TS2742: The inferred type of 'PartialNullable' cannot be named without a reference to './node_modules/@sinclair/typebox/build/cjs/type/computed'. This is likely not portable. A type annotation is necessary.
index.ts(11,17): error TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.

I cannot seem to find an explicit type annotation that will make the compiler happy.

---

Is there a breaking change in the 0.34.1 update? Or is there a better way to do this?

---

Thanks for everything.

**Comment by sinclairzx81** [comment]
@mgreystone Hi, thanks for reporting.

There were some updates on 0.34.1 related to both KeyOf() and Index(). I'll take a look at this and report back.

Cheers

**Comment by sinclairzx81** [comment]
@mgreystone Hiya,

Have taken a quick look at this and the issue relates to recent updates to support deferred computed types (this went out on 0.34.1). I do consider this to be a bug, so will mark it as such, but may not be able to provide you an immediate fix (will try resolve later on this week). 

> Note: You can work around the issue by not exporting the PartialNullable function.

The issue specifically has to do with attempts  to try and clean up the Type.Index signature which had grown exceedingly complicated over the past 12 months. I had implemented a fairly sophisticated conditional mapped type to handle various overloads for Type.Index, but unfortunately, the conditional logic is causing .d.ts generation to blow out.

The fix will likely be to revert the explicit overloaded signatures, but it will take some time to implement while retaining the new TComputed functionality. I am currently on contract work atm, but will try to have this sorted as soon as I can.

Thanks again for reporting.
Cheers
S

**Comment by sinclairzx81** [comment]
@mgreystone Hi, 

This issue should be resolved on 0.34.10. Should be able to test on the TSP link below.

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo43RlAZQGMALDEAQxz0xQHkAjAKw2PgF84AzKCEOAHIAAgGdgAO2IAbSsCgB6WhkYQAHoIBQm7gFcpMYBAlwAcrumzG0jAB5kcDGpgYJAE1FEyFSgD4AFMgAapTSAFwoAJSImnBwUBgwulAmaJgAdACqEkYS-gDawaE4aRjp5pb+kQC6kZoc2noGuXD0YIbGoRVWNvaOzq4eXuRUAUXhUTFxCUkpKPjpbR0Sof7dlNYYgSHSkZEA3PXaTpCwPPrsLQAKlLDAXRY9dg5OLu6eyEys7GNfEcjRBCxeKJZKpBYAWUoYEwbkCCwA0hhUPRuIEvpEcP4EdEALy+YFxJa5B6WDY2eEZACS7ic6JYOBxmOBB3qQA)

```typescript
import { Type, type TSchema, type TObject } from '@sinclair/typebox'

function Nullable<T extends TSchema>(TVal: T) {
  return Type.Union([TVal, Type.Null()])
}

function OptionalNullable<T extends TSchema>(TVal: T) {
  return Type.Optional(Nullable(TVal));
}

export function PartialNullable<T extends TObject>(TObj: T) {
  return Type.Mapped(Type.KeyOf(TObj), (K) =>
    OptionalNullable(Type.Index(TObj, K)),
  );
}
```

---

I note these TS declaration emit type issues tend to be some of the most challenging to track down and solve. Your repro PartialNullable type effectively passed all runtime and static types, but only fails when exporting the type across module boundaries. I will need to spend some time working on a system to catch these kind of problems (writing inference types that are able to synthesize as d.ts definitions is a bit of dark art (especially for some of these more sophisticated types like Mapped and Index).

Anyway, thanks again for reporting. If you run into any problems, feel free to ping on this thread.
All the best
S

**Comment by mgreystone** [comment]
Confirmed working on my end. @sinclairzx81 thank you so much for everything.
